### PR TITLE
SAM-3342 Handling couple of exceptions for TimedAssessmentRunnable th…

### PR DESCRIPTION
**Pull request by yuanhuaqu**
…read

### CODE REVIEW

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/queue/delivery/TimedAssessmentRunnable.java [hunk 1]
```diff
@@ -171,7 +196,13 @@ public void run(){
                true,
                SamigoConstants.NOTI_EVENT_ASSESSMENT_TIMED_SUBMITTED));
 
-            notifyGradebookByScoringType(ag, timedAG.getPublishedAssessment());
+            try {
+               notifyGradebookByScoringType(ag, timedAG.getPublishedAssessment());
+            } catch (GradebookServiceException e) {
+               e.printStackTrace();
```
Comment (ottenhoff) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/queue/delivery/TimedAssessmentRunnable.java]: Should log instead of printing a stack

Comment (ottenhoff) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/queue/delivery/TimedAssessmentRunnable.java]: @yuanhuaqu can you fix?

**Comment by ottenhoff** [comment]
@dnhutchins can you review?

**Comment by dnhutchins** [comment]
This would also allow the ""SAMIGO_TIMED_ASSESSMENT:SUBMIT:FORGRADE assessmentId:" log to be reached, which is something that we don't want to happen, as that would indicate a successful grade submission. 

threadPool.setRemoveOnCancelPolicy(true); is being set in the TimedAssessmentQueue, so if an exception is encountered with notifyGradebookByScoringType, or anywhere else in the runnable for that matter, the     

```
   } catch (Exception ex) {
      log.error("SAMIGO_TIMED_ASSESSMENT:SUBMIT:ERROR - " + ex);
      this.queue.remove(this.timedAGId);
    }
```

catchall, will log it, then remove the task from the queue. and if it was the last task in, the threadPool will remove the thread all together.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/queue/delivery/TimedAssessmentRunnable.java (hunk 1 above)
Comment (dnhutchins) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/queue/delivery/TimedAssessmentRunnable.java]: This is all redundant, that is all inside one big try/catch, on purpose, and we want the failure to log with a failure message, if this exception is eaten here, the assessment is still not going to be submitted, it's just going to continue and log a message that it was successful.

**Comment by yuanhuaqu** [comment]
@dnhutchins @ottenhoff Thanks for reviewing the patch.
@dnhutchins I got your point. Yes, the big block of try/catch under the run() will handle all exceptions and won't suspend the thread (unlike 10.7 and 11 where there is no try/catch).  But also, the auto submission will fail (let's just assume no client side triggering).

As for version 12 and beyond, the focus will be fixing the **auto test submission**.  Therefore, the first fixed part is to make sure that timed test will be submitted with this special case where no event log recorded.  I believe this is needed to be fair for all students for auto-test submission;  should not fail in any case (even with try/catch); with this in mind,  it needs the second part for try/catch the exception of gradebookservice to move on for submission queue remove.  We could afford for gradebookservice exception, but not for auto-submission failure.

**Comment by dnhutchins** [comment]
I'm not sure what you mean by "auto test submission" needing fixing. 

I'm trying to understand if there is something that needs changing here, so bear with me, here are my thoughts.

In order for the this proposed change to be reached, the the conditionals near the top would need to have evaluated true

```
      if (!submitted){
        if (currentTime > bufferedExpirationTime){ // time's up, i.e. timeLeft + latency buffer reached
```
Where the first call made, is to:

```
          timedAG.setSubmittedForGrade(true);
```

Which means that the assessment will already be submitted (as far as samigo is concerned).

with the proposed changes, if  `GradebookServiceException` is thrown trying to send to GB, the next code run will be the logging of a success message, and clearing of the thread's `UsageSession`:

```
            log.info("SAMIGO_TIMED_ASSESSMENT:SUBMIT:FORGRADE assessmentId:" + eventLogData.getAssessmentId() + 
               " userEid:" + eventLogData.getUserEid() + 
               " siteId:" + siteId + 
               " submissionId:" + ag.getAssessmentGradingId());
            //Invalidate the session
            UsageSessionService.logout();
```
where the runnable will terminate, to be re-run on the next pass of the scheduled task, in the next execution of the runnable, "submitted" will be true, so the conditional will continue to:

```
      } else { //submitted, remove from queue if transaction buffer is also reached
        if (currentTime > (bufferedExpirationTime + timedAG.getTransactionBuffer()*1000L)){
          this.queue.remove(this.timedAGId);
        }
      }
```

Now, as the code is now, without the proposed changes, the `GradebookServiceException` would be thrown, skipping the first block above, and executing the following:

```
    } catch (Exception ex) {
      log.error("SAMIGO_TIMED_ASSESSMENT:SUBMIT:ERROR - " + ex);
      this.queue.remove(this.timedAGId);
    }
```

In both cases the task is removed from the thread with `this.queue.remove(this.timedAGId);`
So the only additional code executed with this would be

Logging that we would not want to be associated with a failed submission:
```
            log.info("SAMIGO_TIMED_ASSESSMENT:SUBMIT:FORGRADE assessmentId:" + eventLogData.getAssessmentId() + 
               " userEid:" + eventLogData.getUserEid() + 
               " siteId:" + siteId + 
               " submissionId:" + ag.getAssessmentGradingId());
```

and the closing of the `UsageSession` created as part of the submission:
```
            //Invalidate the session
            UsageSessionService.logout();
```


This is all to try to understand, what is the issue we are trying to fix here? Maybe there is something needed, but I'm not sure I understand what it is we are trying to fix.

Thanks:
  -David

**Comment by yuanhuaqu** [comment]
@dnhutchins You are absolutely right. My concern seems not an issue anymore after I reexamined the code and difference before sakai 11.4-  and 12+.  Thanks a lot.

**Pull request closed without merging**